### PR TITLE
Struct Editor: Implement javascript editor framework 

### PR DESCRIPTION
**Pull request by fpichler**
https://synjira.atlassian.net/browse/CELDEV-426

### CODE REVIEW

**Comment by fpichler** [comment]
@ebeutler, @dkhasham  refoactoring is finished. Please review this PR.

File: web-module/src/main/webapp/resources/structEditJS/StructEditor.js [hunk 1]
```diff
@@ -0,0 +1,801 @@
...
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+/*
+*
+*
+**/
```
Comment (ebeutler) [inline web-module/src/main/webapp/resources/structEditJS/StructEditor.js]: empty comment. Either add comment or remove.

File: web-module/src/main/webapp/resources/structEditJS/StructEditor.js [hunk 2]
```diff
@@ -0,0 +1,801 @@
...
+(function(window, undefined) {
+  "use strict";
+
+  if(typeof window.CELEMENTS.structEdit=="undefined"){window.CELEMENTS.structEdit={};};
+  if(typeof window.CELEMENTS.Utils=="undefined"){window.CELEMENTS.Utils={};};
```
Comment (ebeutler) [inline web-module/src/main/webapp/resources/structEditJS/StructEditor.js]: structEdit lower case, Utils upper case -> do we (or should we?) have come naming conventions for these?

File: web-module/src/main/webapp/resources/structEditJS/StructEditor.js [hunk 3]
```diff
@@ -0,0 +1,801 @@
...
+        if (entry.value.isDirty()) {
+          dirtyEditors.set(entry.key, entry.value);
+        }
+      });
+      return dirtyEditors;
+    },
+
+    hasDirtyEditors : function() {
+      var _me = this;
+      //TODO add possibility to add JS-listener which can do additional 'isDirty' checks
```
Comment (ebeutler) [inline web-module/src/main/webapp/resources/structEditJS/StructEditor.js]: Is there an issue for this? If so, please link in comment

File: web-module/src/main/webapp/resources/structEditJS/StructEditor.js [hunk 4]
```diff
@@ -0,0 +1,801 @@
...
+
+    execShowProgressDialog : function() {
+      var _me = this;
+      var savingDialog = _me._getModalDialog();
+      savingDialog.setHeader(window.celMessages.structEditor.savingDialogHeader);
+      savingDialog.setBody(_me._loading.getLoadingIndicator(true));
+      savingDialog.cfg.queueProperty("buttons", null);
+      savingDialog.render();
+      savingDialog.show();
+      //TODO add possibility to add JS-listener which can execute alternative save actions
```
Comment (ebeutler) [inline web-module/src/main/webapp/resources/structEditJS/StructEditor.js]: Is there an issue for this? If so, please link in comment

File: web-module/src/main/webapp/resources/structEditJS/StructEditor.js [hunk 5]
```diff
@@ -0,0 +1,801 @@
...
+     *
+     * @param jsonResponses
+     * @returns true if errors have been displayed
+     *          false if no errors have been displayed
+     */
+    _showErrorMessages : function(jsonResponses) {
+      var _me = this;
+      var errorMessages = new Array();
+      jsonResponses.each(function(response) {
+  //            var formId = response.key;
```
Comment (ebeutler) [inline web-module/src/main/webapp/resources/structEditJS/StructEditor.js]: needed as comment?

File: web-module/src/main/webapp/resources/structEditJS/StructEditor.js [hunk 6]
```diff
@@ -0,0 +1,801 @@
...
+    _isDirtyField : function(fieldElem) {
+      var _me = this;
+      if (fieldElem.hasClassName('celDirtyOnLoad')) {
+        return true;
+      }
+      var formId = fieldElem.up('form').id;
+      if (fieldElem.hasClassName('mceEditor') && tinyMCE && tinyMCE.get(fieldElem.id)) {
+        //FIXME sometimes isDirty from tinyMCE is wrong... thus we compare the .getContent
+        //FIXME with the _initialValues instead.
+//        return tinyMCE.get(fieldElem.id).isDirty();
```
Comment (ebeutler) [inline web-module/src/main/webapp/resources/structEditJS/StructEditor.js]: Is there an issue for this? If so, please link in comment

File: web-module/src/main/webapp/resources/structEditJS/StructEditor.js [hunk 7]
```diff
@@ -0,0 +1,801 @@
...
+       }
+       return false;
+     },
+
+    saveAllForms : function(allDirtyFormIds) {
+      var _me = this;
+      if (allDirtyFormIds.size() > 0) {
+        var formId = allDirtyFormIds.pop();
+        var remainingDirtyFormIds = allDirtyFormIds;
+        _me._saveAndContinueAjax(formId, {
```
Comment (ebeutler) [inline web-module/src/main/webapp/resources/structEditJS/StructEditor.js]: Either use onComplete and handle problems or add onFailed(?) ... otherwiese e.b. Velocity-Exceptions could lead to wrong behaviours

File: web-module/src/main/webapp/resources/structEditJS/StructEditor.js [hunk 8]
```diff
@@ -0,0 +1,801 @@
...
+        console.log('save done.');
+        _me._saveCallback(_me._jsonResponses.toObject());
+      }
+    }
+
+  });
+  CELEMENTS.structEdit.CelementsFormSaver.prototype = Object.extend(
+      CELEMENTS.structEdit.CelementsFormSaver.prototype, CELEMENTS.mixins.Observable);
+
+})(window);
```
Comment (ebeutler) [inline web-module/src/main/webapp/resources/structEditJS/StructEditor.js]: missing empty line as EOF

File: web-module/src/main/webapp/resources/structEditJS/initStructEditor.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  };
+
+  celAddOnBeforeLoadListener(function() {
+    window.celStructEditorManager = new CELEMENTS.structEdit.StructEditorManager();
+    $(document.body).stopObserving("celements:contentChanged", initStructEditorContentChangedHandler);
+    $(document.body).observe("celements:contentChanged", initStructEditorContentChangedHandler);
+    window.celStructEditorManager.startEditorManager();
+  });
+
+})(window);
```
Comment (ebeutler) [inline web-module/src/main/webapp/resources/structEditJS/initStructEditor.js]: missing empty line as EOF

File: web-module/src/main/webapp/templates/celAjax/Messages.vm [hunk 10]
```diff
@@ -0,0 +1,6 @@
+#set($jsonBuilder = $services.celementsweb.getNewJSONBuilder())
+$jsonBuilder.openDictionary()
+#parse('celMacros/CelMessages.vm')
+#parse('celMacros/StructEditorMessages.vm')
+$jsonBuilder.closeDictionary()
+$jsonBuilder.getJSON()
```
Comment (ebeutler) [inline web-module/src/main/webapp/templates/celAjax/Messages.vm]: missing empty line as EOF

File: web-module/src/main/webapp/templates/celMacros/StructEditorMessages.vm [hunk 11]
```diff
@@ -0,0 +1,10 @@
+$jsonBuilder.openProperty('structEditor')
+$jsonBuilder.openDictionary()
+$jsonBuilder.addStringProperty('unsavedChangesOnCloseMessage', "$!adminMsg.get('cel_tm_unsavedChangesOnCloseMessage')")
+$jsonBuilder.addStringProperty('savingDialogWarningHeader', "$!adminMsg.get('cel_tm_saving_dialog_warning_header')")
+$jsonBuilder.addStringProperty('savingDialogMessage', "$!adminMsg.get('cel_tm_saving_dialog_unsaved_changes')")
+$jsonBuilder.addStringProperty('savingDialogButtonCancel', "$!adminMsg.get('cel_cancel')")
+$jsonBuilder.addStringProperty('savingDialogButtonSave', "$!adminMsg.get('cel_tm_saving_dialog_button_save')")
+$jsonBuilder.addStringProperty('savingDialogButtonDoNotSave', "$!adminMsg.get('cel_tm_saving_dialog_button_donotsave')")
+$jsonBuilder.addStringProperty('savingDialogHeader', "$!adminMsg.get('cel_tm_saving_dialog_header')")
+$jsonBuilder.closeDictionary()
```
Comment (ebeutler) [inline web-module/src/main/webapp/templates/celMacros/StructEditorMessages.vm]: missing empty line as EOF

**Review by ebeutler — CHANGES_REQUESTED** [review]

**Comment by fpichler** [comment]
can you please re-review?

**Review by ebeutler — APPROVED** [review]

**Pull request merged**